feat(reader): Add PartitionSpec support to FileScanTask and RecordBatchTransformer

mbutrovich opened the pull request:
## Which issue does this PR close?

Partially address #1749.

## What changes are included in this PR?

This PR adds partition spec handling to `FileScanTask` and `RecordBatchTransformer` to correctly implement the Iceberg spec's "Column Projection" rules for fields "not present" in data files.

### Problem Statement

Prior to this PR, `iceberg-rust`'s `FileScanTask` had no mechanism to pass partition information to `RecordBatchTransformer`, causing two issues:

1. **Incorrect handling of bucket partitioning**: Couldn't distinguish identity transforms (which should use partition metadata constants) from non-identity transforms like bucket/truncate/year/month (which must read from data file). For example, `bucket(4, id)` stores
`id_bucket = 2` (bucket number) in partition metadata, but actual `id` values (100, 200, 300) are only in the data file. iceberg-rust was incorrectly treating bucket-partitioned source columns as constants, breaking runtime filtering and returning incorrect query results.

2. **Field ID conflicts in add_files scenarios**: When importing Hive tables via `add_files`, partition columns could have field IDs conflicting with Parquet data columns. Example: Parquet has field_id=1→"name", but Iceberg expects field_id=1→"id" (partition). Per spec, the
correct field is "not present" and requires name mapping fallback.

### Iceberg Specification Requirements

Per the Iceberg spec (https://iceberg.apache.org/spec/#column-projection), when a field ID is "not present" in a data file, it must be resolved using these rules:

1. Return the value from partition metadata if an **Identity Transform** exists
2. Use `schema.name-mapping.default` metadata to map field id to columns without field id
3. Return the default value if it has a defined `initial-default`
4. Return null in all other cases

**Why this matters:**
- **Identity transforms** (e.g., `identity(dept)`) store actual column values in partition metadata that can be used as constants without reading the data file
- **Non-identity transforms** (e.g., `bucket(4, id)`, `day(timestamp)`) store transformed values in partition metadata (e.g., bucket number 2, not the actual `id` values 100, 200, 300) and must read source columns from the data file

### Changes Made

1. **Added partition fields to `FileScanTask`** (`scan/task.rs`):
- `partition: Option<Struct>` - Partition data from manifest entry
- `partition_spec: Option<Arc<PartitionSpec>>` - For transform-aware constant detection
- `name_mapping: Option<Arc<NameMapping>>` - Name mapping from table metadata

2. **Implemented `constants_map()` function** (`arrow/record_batch_transformer.rs`):
- Replicates Java's `PartitionUtil.constantsMap()` behavior
- Only includes fields where transform is `Transform::Identity`
- Used to determine which fields use partition metadata constants vs. reading from data files

3. **Enhanced `RecordBatchTransformer`** (`arrow/record_batch_transformer.rs`):
- Added `build_with_partition_data()` method to accept partition spec, partition data, and name mapping
- Implements all 4 spec rules for column resolution with identity-transform awareness
- Detects field ID conflicts by verifying both field ID AND name match
- Falls back to name mapping when field IDs are missing/conflicting (spec rule #2)

4. **Updated `ArrowReader`** (`arrow/reader.rs`):
- Uses `build_with_partition_data()` when partition information is available
- Falls back to `build()` when not available

5. **Updated manifest entry processing** (`scan/context.rs`):
- Populates partition fields in `FileScanTask` from manifest entry data

### Tests Added

1. **`bucket_partitioning_reads_source_column_from_file`** - Verifies that bucket-partitioned source columns are read from data files (not treated as constants from partition metadata)

2. **`identity_partition_uses_constant_from_metadata`** - Verifies that identity-transformed fields correctly use partition metadata constants

3. **`test_bucket_partitioning_with_renamed_source_column`** - Verifies field-ID-based mapping works despite column rename

4. **`add_files_partition_columns_without_field_ids`** - Verifies name mapping resolution for Hive table imports without field IDs (spec rule #2)

5. **`add_files_with_true_field_id_conflict`** - Verifies correct field ID conflict detection with name mapping fallback (spec rule #2)

6. **`test_all_four_spec_rules`** - Integration test verifying all 4 spec rules work together

## Are these changes tested?

Yes, there are 6 new unit tests covering all 4 Iceberg spec rules. This also resolved approximately 50 Iceberg Java tests when running with DataFusion Comet's experimental https://github.com/apache/datafusion-comet/pull/2528 PR.

liurenjie1024 commented on crates/iceberg/src/scan/task.rs:
```diff
@@ -54,6 +58,24 @@ pub struct FileScanTask {
+    /// partition spec and determine which fields are identity-transformed.
+    /// Not serialized as partition data is runtime-only and populated from manifest entries.
+    #[serde(skip)]
+    pub partition_spec_id: Option<i32>,
```
Not needed, we have a `spec_id` in `PartitionSpec`.

liurenjie1024 commented on crates/iceberg/src/scan/task.rs:
```diff
@@ -54,6 +58,24 @@ pub struct FileScanTask {
+    /// constant values from partition metadata vs. reading from the data file.
+    /// Per the Iceberg spec, only identity-transformed partition fields should use constants.
+    #[serde(skip)]
+    pub partition: Option<Struct>,
```
I'm fine with implementing serialization for `Struct` in follow up pr, but I don't think the default behavior is skipping, which may lead to bugs. We could provide a function with [serialize_with](https://serde.rs/field-attrs.html#serialize_with) attribute, and this function could just return an NotImplemented error for now.

liurenjie1024 commented on crates/iceberg/src/arrow/record_batch_transformer.rs:
```diff
@@ -270,57 +394,199 @@ impl RecordBatchTransformer {
+                std::collections::HashSet::new()
+            };
+
+        let has_field_id_conflict = projected_iceberg_field_ids.iter().any(|field_id| {
```
I'm confused about this part. I think the rule about missing column is quite clear in [iceberg spec](https://iceberg.apache.org/spec/#column-projection:~:text=Values%20for%20field%20ids%20which%20are%20not%20present%20in%20a%20data%20file%20must%20be%20resolved%20according%20the%20following%20rules%3A), it's just a fallback chain to lookup.

liurenjie1024 commented on crates/iceberg/src/arrow/record_batch_transformer.rs:
```diff
@@ -270,57 +394,199 @@ impl RecordBatchTransformer {
+        });
+
+        // Build name-based mapping if there's a field ID conflict
+        let name_to_source_schema_map: HashMap<String, (FieldRef, usize)> = source_schema
```
This is incorrect, the iceberg spec says we should use default name mapping in table metadata, rather this way.

mbutrovich replied:
Makes sense. Let me see how bad this would be to implement in this PR or if it start to fall out of scope. Thanks for the clarification!

mbutrovich commented:
Hopefully I addressed all of your comments @liurenjie1024. The `serialize_with` stuff was new to me, so let me know if that's not what you had in mind. Thanks for your patience!

liurenjie1024 commented:
> Hopefully I addressed all of your comments @liurenjie1024. The `serialize_with` stuff was new to me, so let me know if that's not what you had in mind. Thanks for your patience!

Thanks, I'll take a look today.

liurenjie1024 reviewed (commented):
Thanks @mbutrovich for this pr! I think mostly LGTM, we just need adjustment according to the rule

mbutrovich commented:
> Thanks @mbutrovich for this pr! I think mostly LGTM, we just need adjustment according to the rule

I'll take another pass today, thanks for the further comments! It's a non-trivial change and I'm still learning my way around the codebase and spec, so I appreciate your patience.

liurenjie1024 commented on crates/iceberg/src/arrow/record_batch_transformer.rs:
```diff
@@ -123,12 +166,55 @@ impl RecordBatchTransformer {
+    /// - Spec: https://iceberg.apache.org/spec/#column-projection
+    /// - Java: core/src/main/java/org/apache/iceberg/util/PartitionUtil.java
+    /// - Java: parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java
+    pub(crate) fn build_with_partition_data(
```
nit: How about we creat a builder struct for `RecordBatchTransformer`?

mbutrovich replied:
Updated.

liurenjie1024 reviewed (commented):
Thanks @mbutrovich for this explaination. I still didn't find related code in java repo, could you help to point out?

mbutrovich commented on crates/iceberg/src/arrow/record_batch_transformer.rs:
```diff
@@ -29,9 +29,45 @@ use arrow_schema::{
+        // Only identity transforms should use constant values from partition metadata
+        if matches!(field.transform, Transform::Identity) {
+            // Get the partition value for this field
+            if let Some(Some(Literal::Primitive(value))) = partition_data.iter().nth(pos) {
```
Updated.

mbutrovich commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -919,6 +954,77 @@ fn build_fallback_field_id_map(parquet_schema: &SchemaDescriptor) -> HashMap<i32
+///
+/// # Returns
+/// Arrow schema with field IDs assigned based on name mapping
+fn apply_name_mapping_to_arrow_schema(
```
Opened https://github.com/apache/iceberg-rust/issues/1845
